zcash_client_sqlite/src/wallet/commitment_tree.rs

```diff
@@ -367,6 +387,46 @@ pub(crate) fn last_shard<H: HashSer>(
     .transpose()
 }
 
+/// Returns an error iff the proposed insertion range
+/// for the tree shards would create a discontinuity
+/// in the database.
+fn check_shard_discontinuity(
+    conn: &rusqlite::Connection,
+    table_prefix: &'static str,
+    proposed_insertion_range: Range<u64>,
+) -> Result<(), Error> {
+    if let Ok((Some(min), Some(max))) = conn
+        .query_row(
+            &format!(
+                "SELECT MIN(shard_index), MAX(shard_index) FROM {}_tree_shards",
+                table_prefix
+            ),
+            [],
+            |row| {
+                let min = row.get::<_, Option<u64>>(0)?;
+                let max = row.get::<_, Option<u64>>(1)?;
+                Ok((min, max))
+            },
+        )
+        .map_err(Error::Query)
+    {
+        if !proposed_insertion_range.contains(&min) && !proposed_insertion_range.contains(&max) {
```

Comment:
There was a bug here, that my nit was pointing towards (but I reviewed this at midnight so I didn't spot it then): This checks for the proposed insertion range containing the existing range, but not for the reverse (which is far more common). We need to actually check that the ranges do not overlap before proceeding to the adjacency check below.